### DISCUSSION TITLE
The evolution of cooperation

### DISCUSSION THREAD

**Discussion by ta-rekt** [josePereiro/TankeFactory#11] [Discussions]
> "To see what type of strategy can thrive in a variegated environment of more or less sophisticated strategies, one of us (R.A.) conducted a computer tournament for the Prisoner's Dilemma. The strategies were submitted by game theorists in economics, sociology, political science, and mathematics (22). The rules implied the payoff matrix shown in Fig. I and a game length of 200 moves. The 14 entries and a totally random strategy were paired with each other in a round robin tournament. Some of the strategies were quite intricate. An example is one which on each move models the behavior of the other player as a Markov process, and then uses Bayesian inference to select what seems the best choice for the long run. However, the result of the tournament was that the highest average score was attained by the simplest of all strategies submitted: TIT FOR TAT. This strategy is simply one of cooperating on the first move and then doing whatever the other player did on the preceding move. Thus TIT FOR TAT is a strategy of cooperation based on reciprocity. The results of the first round were then circulated and entries for a second round were solicited. This time there were 62 entries from six countries (23). Most of the contestants were computer hobbyists, but there were also professors of evolutionary biology, physics, and computer science, as well as the five disciplines represented in the first round. TIT FOR TAT was again submitted by the 27 MARCH 1981 winner of the first round, Professor Anatol Rapoport of the Institute for Advanced Study (Vienna). It won again. An analysis of the 3 million choices which were made in the second round identified the impressive robustness of TIT FOR TAT as dependent on three features: it was never the first to defect, it was provocable into retaliation by a defection of the other, and it was forgiving after just one act of retaliation (24). The robustness of TIT FOR TAT was also manifest in an ecological analysis of a whole series of future tournaments. The ecological approach takes as given the varieties which are present and investigates how they do over time when interacting with each other. This analysis was based on what would happen if each of the strategies in the second round were submitted to a hypothetical next round in proportion to its success in the previous round. The process was then repeated to generate the time path of the distribution of strategies. The results showed that, as the less successful rules were displaced, TIT FOR TAT continued to do well with the rules which initially scored near the top. In the long run, TIT FOR TAT displaced all the other rules and went to fixation (24). This provides further evidence that TIT FOR TAT'S cooperation based on reci- procity is a robust strategy that can thrive in a variegated environment"

From this paper:
[https://websites.umich.edu/~axe/research/Axelrod%20and%20Hamilton%20EC%201981.pdf](url)

This is part of the argument that cooperative strategies can emerge and remain stable in a population of many strategies, including nasty ones. In the next paragraph he proves that TIT FOR TAT is evolutionarily stable (ES) with a really simple argument. He argues that even if the population is dominated by the ultimate nasty strategy that always defects, which is also ES, if TIT FOR TAT appears in a cluser and starts growing, it can displace the nasty strategy and be resistant. This is counter-intuitive, since the definition of ES is that a strategy can't be displaced by any mutant. Now it is a question of thesholds, above which it is possible to displace and below which it is not.

This is a path to attack the question you asked when we were walking to the Eiffel tower: how can we know for sure that a species will fixate, and how long does it take?

There's many things I would like to try with this. One is trying to implement strategies as a code that can be mutated, and ask how long it would take evolution to discover a strategy like TIT FOR TAT by random mutations. In this paper they incorporate only one side of evolution, which is selection. They start with hard-coded strategies and see which get selected and which go extinct. I would like to try the mutation aspect as well: how can they emerge on their own, and is it tractable to try this out.

Just discussing here. What are your thoughts?

**Comment by josePereiro** [comment]
That is great! I propose the follow (I'm sorry that it is mostly work for you):
1. (ta-rekt) Create a github repo for working in the code together. I propose `python` to be able to scale to pro `ML` libraries quickly (like `keras`, `pytorch`, `tensorflow`, etc...)...
   1.1. (ta-rekt) Program a quick example as the seed (ex: see below)...
2. (ta-rekt) Recommend reading list
3. (jose) read and start contributing to the code
4. (everybody) keep in touch 

### Additional Materials

Here are some quick simulations I just did in `julia`, they plot the total sentence (prision years) of different strategies across 1000 matches of 200 trials each... Duplicating this in python could be a good start... After you do the repo I can upload the code for reference...
<img width="561" alt="image" src="https://github.com/josePereiro/TankeFactory/assets/44244693/ec9f3625-b8a7-4cbe-aed7-a3df7c161922">
<img width="561" alt="image" src="https://github.com/josePereiro/TankeFactory/assets/44244693/36b473ab-0625-4b0f-b365-3250ad4b7aa8">

I used this payoff matrix (from [wikipedia](https://en.wikipedia.org/wiki/Prisoner%27s_dilemma)) for each trial:
<img width="376" alt="image" src="https://github.com/josePereiro/TankeFactory/assets/44244693/8b49d9c9-7303-49eb-89fe-dd8a4162cd99">

**Reply by josePereiro** [reply]
Lets just start rolling!!!

**Reply by josePereiro** [reply]
I don't make the repo myself because I don't know how a `python` project is structured. Maybe you have more experience... Otherwise I'll do it.

**Reply by ta-rekt** [reply]
Done. It took a while, but I did some very basic things in a repo called [`evolution-of-cooperation`](https://github.com/ta-rekt/evolution-of-cooperation), to which I invited you. I implemented 1-vs-1 iterated prisoner's dilemma, 6 simple strategies,. I am now getting tournaments to work, so that we can compare a population of agents.

**Reply by ta-rekt** [reply]
There is this repository called [`Axelrod`](https://github.com/Axelrod-Python/Axelrod)

It's well-documented and tested, and people use it to reproduce research in evolutionary game theory. Maybe we could use it to benchmark our results. Although it seems a bit rigid. For example I didn't get a sense that it was easy to program agents to evolve.

Anyway, we can get inspired and learn from their implementation.

**Reply by josePereiro** [reply]
greeeeeaat! 

I'll check all this and give a feedback zoom.

**Reply by josePereiro** [reply]
https://github.com/Axelrod-Python/Axelrod-notebooks, I'll start by this notebooks

**Reply by josePereiro** [reply]
> It's well-documented and tested, and people use it to reproduce research in evolutionary game theory. Maybe we could use it to benchmark our results. Although it seems a bit rigid. For example I didn't get a sense that it was easy to program agents to evolve.

check this https://github.com/Axelrod-Python/axelrod-dojo
